asynciterator.ts

```diff
@@ -1981,42 +1984,170 @@ class HistoryReader<T> {
   }
 }
 
+/**
+ * An iterator that takes a variety of iterable objects as a source.
+ */
+export class WrappingIterator<T> extends AsyncIterator<T> {
+  protected _source: InternalSource<T> | null = null;
+
+  constructor(source?: MaybePromise<IterableSource<T>>) {
+    super();
+    if (source) {
+      if (!isPromise(source))
+        this.source = source as any;
+      else
+        source.then(s => this.source = s).catch(e => this.emit('error', e));
```

Comment:
This has the potential to lead to a gnarly race condition.

In particular if I run
```ts
const iter = new WrappingIterator(new Promise(resolve => {   
   setTimeout(() => { resolve(range(1, 5)) }, 1000)
   }))
iter.source = range(10)
```

Then when we start reading from the iterator it will return `10, 11, ...` for the first second whilst the input source is still resolving; and then it will go `1, 2, 3, 4, 5` before closing.

That is to say, needs to be some sort of wait to invalidate this setting operation if `set source` is called afterwards.

---

The easier option is to make `set source` internal and have this warning about it's usage that links to this comment. Then no code changes are required.

Reply:
Big +1 on this, `set source()` on `WrappingIterator` should not be part of the public API.

Reply:
Ugh and I was so glad I remembered 56fc2b23bb74028cde06ad2704e3257e79b696ae 😅

Absolutely right, we need to guard against pending promises.

> set source() on WrappingIterator should not be part of the public API

Actually there's an interesting use case, which is why I did 98eb9cfb19d4640cdb8d19083e7dedf1758e39b7: that way, you can use `WrappingIterator` as a placeholder and then insert the actual iterator when it is ready. `TransformIterator` supports that same pattern.

So I'll fix the pending promise issue.